File: js-core/Evaluator.js


exports.evaluate = function (exprl, symbolTable, extraFields) {
  var state = EvalState()
  state.symTab = symbolTable
  Object.keys(extraFields).forEach (
    function (k) { state[k] = extraFields[k] }
  )

  exprl.forEach (function (expr) { var r = state.evaluate (expr, symbolTable)
                                   state.values.push (r) 
                                 }
                )

  return state
}


function EvalState () {
  return {
    values:   [],        // output values from top-level expression list
    evaluate: _evaluate
  }
}


function _evaluate (expr) {
  var f = EvalTable[expr.type]
  if (! f) { throw 'unkown expression type - ' + expr.type }
  return f.apply(this,[expr])
}


var EvalTable = {
  Assignment: function (e) {
    var value = this.evaluate (e.e)
    this.symTab.updateBindingForValue (e.v, value)
    return value
  },
  Symbol: function (e) {
    var binding = this.symTab.retrBinding (e.v)
    return binding.v
  },
  Defun: function (e) {
    // fix: symbol table derivation
    this.symTab.updateBindingForDefun (e.v, e.arity, e.e, null)
    return e.arity
  },
  Funcall: function (e) {
    binding = this.symTab.retrBinding (e.v)
    if (binding.type == 'BoundBuiltin') {
      return binding.f.apply (this, [e.exprl])
    }
    if (binding.type == 'BoundDefun') {
      var args = []
      // FIX: do symbol table stuff for args (and scope)
      for (var i=0; i<binding.arity; i++) {
        args.push (this.evaluate (e.exprl[i]))
      }
      return this.evaluate (binding.e)
    }
  },
  BinaryOp: function (e) {
    var left  = this.evaluate (e.left)
    var right = this.evaluate (e.right)
    return e.f (left, right)
  },
  UnaryOp: function (e) {
    return e.f (this.evaluate(e.e))
  },
  ConstantNumber: function (e) {
    return e.v
  },
  TernaryIf: function (e) {
    var cond = this.evaluate (e.econd)
    if (cond != 0) {
      return this.evaluate (e.eif)
    } else {
      return this.evaluate (e.ethen)
    }
  },
  ExprTreeListNode: function (e) {
    var outerSymTab = this.symTab,
        result
    this.symTab = this.symTab.derivedTable ()
    for (var i=0; i<e.exprl.length; i++) {
      result = this.evaluate (e.exprl[i])
    }
    this.symTab = outerSymTab
    return result
  },
  Repeat: function (e) {
    // FIX: and, again, symbol table ... tricky block difference from
    // singular expression
    var ntimes = this.evaluate (e.ntimes),
        result;
    for (var i=0; i<ntimes; i++) {
      result = this.evaluate (e.e)
    }
    return result;
  }

}

//

function dbg (s) { console.log(s) }

File: hs-src/lib/Evaluator.hs
module Evaluator
       where

import SymbolTable
import ExprTree
import Control.Monad.State.Strict

evaluate :: ExprTree -> EvalContext

evaluate (Assignment sym exprt) = do
  num <- evaluate exprt
  updateBinding sym (BoundValue num)
     
evaluate (Symbol sym) = do
  binding <- getBinding sym
  case binding of
    (BoundValue num) -> return num
    otherwise -> error "shouldn't see other bindings here in Symbol eval def"

evaluate (Defun sym arity exprt) = do
  -- we need to put a binding in place (we only put in placeholders
  -- during parse)
  st <- getST
  updateBinding sym $ BoundDefun arity exprt st
  return $ fromIntegral arity

evaluate (Funcall arity sym args) = do
  binding <- getBinding sym
  case (binding) of
    (BoundBuiltin arityInTable f) -> f args
    (BoundDefun arityInTable exprt defunScopeST) -> do
      -- we want to evaluate the each item in the args list and bind
      -- the results to a..i in a new, local, symbol table.
      argValues <- mapM evaluate args
      let locals = zip ['a'..'i'] (map BoundValue argValues)
      callScopeST <- getST
      putST $ funcScopedTable locals defunScopeST callScopeST
      result <- evaluate exprt
      (SymbolTable _ (Just originalST')) <- getST
      putST originalST'
      return result
      
evaluate (UnaryOp (UnaryFunc _ f) exprt) = do -- liftM f (evaluate exprt)
  v <- evaluate exprt
  v `seq` return $! f v
evaluate (BinaryOp (BinaryFunc _ f) left right) = do
  -- originally was the very satisfying:
  --   liftM2 f (evaluate left) (evaluate right)  
  -- rewritten to force strict evaluation
  l <- evaluate left
  r <- evaluate right
  l `seq` r `seq` return $! f l r

evaluate (ConstantNumber num) = return num

evaluate (TernaryIf eCond eIf eThen) = do
  num <- evaluate eCond
  if num /= 0 then evaluate eIf else evaluate eThen

evaluate (ExprTreeListNode exprl) = do
  originalST <- getST
  putST $ derivedTable [] originalST
  results <- mapM evaluate exprl
  (SymbolTable _ (Just originalST')) <- getST
  putST originalST'
  return $ last results 

evaluate (Repeat exprNumTimes exprt) = do
  numTimes <- evaluate exprNumTimes
  -- need to handle expression list blocks differently from bare
  -- expressions, here for a block, we want to repeat multiple times
  -- with a stateful symbol table across all repeats, then "pop" that
  -- symbol table and throw it away.
  case exprt of
    (ExprTreeListNode exprl) -> do
      originalST <- getST
      putST $ derivedTable [] originalST
      r <- repeat numTimes (last `liftM` (mapM evaluate exprl)) 0
      (SymbolTable _ (Just originalST')) <- getST
      putST $! originalST'
      return $! r
    otherwise -> do
      r <- repeat numTimes (evaluate exprt) 0
      return $! r
  where repeat n f acc | n <= 0    = return acc
                       | otherwise = do
                           r' <- f
                           repeat (n-1) f r'
